### PR TITLE
Impro-1121: Rename probability cubes - 2

### PR DESCRIPTION
**Pull request by markysparks**
Addresses #GitHubissuenum

Rename all IMPROVER cubes that contain probabilities like `probability_of_rainfall_rate` to follow a specified convention that includes whether the probability is above or below a threshold within the cube name, so that the cube name better describes the contents of the cube. For example:

- `probability_of_rainfall_rate_above_threshold`
- `probability_of_visibility_in_air_below_threshold`

This is an additional PR building on the changes made in PR #813 . The changes made here take in to account the special case of snow falling that uses percentiles-to-probabilities. A new (output) name such as `probability_of_snow_level_being_at_surface` must now be provided as a mandatory argument for this CLI.

I have had to merge in the changes from PR #813 to get all the tests to pass when working on this branch, however here is a list of the files that have been changed outside of this merge:

        modified:   lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py
        modified:   lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py
        modified:   lib/improver/tests/utilities/test_ProbabilitiesFromPercentiles2D.py
        modified:   lib/improver/tests/wxcode/wxcode/test_WeatherSymbols.py
        modified:   lib/improver/utilities/statistical_operations.py
        modified:   tests/improver-percentiles-to-probabilities/00-null.bats
        modified:   tests/improver-percentiles-to-probabilities/01-help.bats
        modified:   tests/improver-percentiles-to-probabilities/02-basic.bats
        modified:   tests/improver-percentiles-to-probabilities/03-rename.bats

Testing:
 - [x] Ran tests and they passed OK
 - [x] Updated tests as required by changes

### CODE REVIEW

File: bin/improver-percentiles-to-probabilities [hunk 1]
```diff
@@ -78,17 +78,17 @@ def main():
...
                         help="The output path for the processed NetCDF")
-    parser.add_argument("--new_name", metavar="NEW_NAME", default=None,
-                        type=str,
-                        help="Name for data in output file.  Defaults to "
-                        "'probability_of_X', where X is the name of the "
-                        "percentiled diagnostic.")
+    parser.add_argument("output_diagnostic_name",
+                        metavar="OUTPUT_DIAGNOSTIC_NAME", type=str,
+                        help="Name for data in output file e.g. "
+                             "probability_of_snowfall")
```
Comment (cgsandford) [inline bin/improver-percentiles-to-probabilities]: This docstring needs to conform to the IMPROVER standard, which was agreed as eg "probability_of_snow_falling_level_below_ground_level".

File: lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py [hunk 2]
```diff
@@ -143,21 +143,13 @@ def _add_bounds_to_percentiles_and_forecast_at_percentiles(
...
-                **forecast_at_percentiles** (Numpy array):
-                    Array containing the underlying forecast values at each
-                    percentile.
+            Array of percentiles from a Cumulative Distribution Function.
+        forecast_at_percentiles (Numpy array):
+            Array containing the underlying forecast values at each
+            percentile.
+        bounds_pairing (Tuple):
+            Lower and upper bound to be used as the ends of the
+            cumulative distribution function.
```
Comment (cgsandford) [inline lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py]: This docstring change looks unintentional - please would you revert this.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py [hunk 3]
```diff
@@ -44,16 +43,18 @@
 from improver.utilities.cube_manipulation import concatenate_cubes
 
 
-def set_up_probability_above_threshold_cube(
+def set_up_probability_threshold_cube(
         data, phenomenon_standard_name, phenomenon_units,
         forecast_thresholds=np.array([8, 10, 12]), timesteps=1,
-        y_dimension_length=3, x_dimension_length=3):
+        y_dimension_length=3, x_dimension_length=3,
+        relative_to_threshold='above'):
```
Comment (cgsandford) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py]: The changes to generalise this aren't necessary - there is work going on in #742 to replace these set up functions.  Adding an optional argument `relative_to_threshold` is OK but please would you revert the function name, as this creates a lot of extra lines of code to review for no reason.

File: lib/improver/tests/utilities/test_ProbabilitiesFromPercentiles2D.py [hunk 4]
```diff
@@ -103,17 +103,19 @@ def test_basic(self):
...
-        """ Test default naming """
-        plugin_instance = ProbabilitiesFromPercentiles2D(self.test_cube)
-        self.assertEqual(plugin_instance.output_name,
-                         "probability_of_{}".format(self.test_cube.name()))
+    # def test_naming(self):
+    #     """ Test default naming """
+    #     plugin_instance = ProbabilitiesFromPercentiles2D(self.test_cube)
+    #     self.assertEqual(
+    #         plugin_instance.output_name,
+    #         "probability_of_{}_above_threshold".format(self.test_cube.name()))
```
Comment (cgsandford) [inline lib/improver/tests/utilities/test_ProbabilitiesFromPercentiles2D.py]: Please remove commented-out block

File: tests/improver-percentiles-to-probabilities/02-basic.bats [hunk 5]
```diff
@@ -38,7 +38,7 @@
   run improver percentiles-to-probabilities \
       "$IMPROVER_ACC_TEST_DIR/percentiles-to-probabilities/snow_level.nc" \
       "$IMPROVER_ACC_TEST_DIR/percentiles-to-probabilities/enukx_orography.nc" \
-      "$TEST_DIR/output.nc"
+      "$TEST_DIR/output.nc" "probability_of_falling_snow_level_asl"
```
Comment (cgsandford) [inline tests/improver-percentiles-to-probabilities/02-basic.bats]: This cube name is not correct or to the standard that was agreed.  Please could you change this to "probability_of_snow_falling_level_below_ground_level".

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py (hunk 3 above)
Comment (bayliffe) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py]: I don't quite follow. The reason the name has been changed is that this function no longer creates a probability cube above a threshold, it created a cube that is relative to the threshold as defined by the relative_to_threshold argument. If I change the name back we will have a function with a name that does not describe what it is doing.

The addition of this keyword was made to remove a copy/paste of this complete function that had been introduced to create cubes below a threshold.

File: lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py (hunk 2 above)
Comment (bayliffe) [reply lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py]: Good spot. Done.

File: bin/improver-percentiles-to-probabilities (hunk 1 above)
Comment (bayliffe) [reply bin/improver-percentiles-to-probabilities]: Done.

File: lib/improver/tests/utilities/test_ProbabilitiesFromPercentiles2D.py (hunk 4 above)
Comment (bayliffe) [reply lib/improver/tests/utilities/test_ProbabilitiesFromPercentiles2D.py]: Done.

File: tests/improver-percentiles-to-probabilities/02-basic.bats (hunk 5 above)
Comment (bayliffe) [reply tests/improver-percentiles-to-probabilities/02-basic.bats]: I have updated the name here and in the corresponding KGO.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py (hunk 3 above)
Comment (cgsandford) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py]: OK.

**Pull request merged**